### PR TITLE
#356 test and partial fix

### PR DESCRIPTION
**Pull request by lambdaupb**
This hopefully fixes the `ClasspathResourceHandler` (and with it `PublicResourceHandler`) while the `WebjarsResourceHandler` is still insecure.

A test is implemented that checks some of the resource handlers with a naive case of this issue.

see #356

### CODE REVIEW

File: pippo-core/src/main/java/ro/pippo/core/route/ClasspathResourceHandler.java [hunk 1]
```diff
@@ -15,31 +15,41 @@
...
 
-        this.resourceBasePath = getNormalizedPath(resourceBasePath);
+        this.resourceBasePath = Paths.get(getNormalizedPath(resourceBasePath));
     }
 
     @Override
     public URL getResourceUrl(String resourcePath) {
-        String resourceName = getResourceBasePath() + "/" + resourcePath;
-        return this.getClass().getClassLoader().getResource(resourceName);
+        Path resolved = resourceBasePath.resolve(resourcePath).normalize();
```
Comment (decebals) [inline pippo-core/src/main/java/ro/pippo/core/route/ClasspathResourceHandler.java]: I prefer to keep `resourceBasePath` as String and to modify only `getResourceUrl`.
The modification create a `Path requestedPath` variable and check to see if this path startWith resourceBasePath (similar with https://github.com/decebals/pippo/blob/master/pippo-core/src/main/java/ro/pippo/core/route/DirectoryHandler.java#L138)

File: pippo-core/src/test/java/ro/pippo/core/PathTraversalResourceHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package ro.pippo.core;
```
Comment (decebals) [inline pippo-core/src/test/java/ro/pippo/core/PathTraversalResourceHandlerTest.java]: Please add copyright header (see other .java files).

File: pippo-core/src/test/java/ro/pippo/core/PathTraversalResourceHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+public class PathTraversalResourceHandlerTest {
+
+  @Test
```
Comment (decebals) [inline pippo-core/src/test/java/ro/pippo/core/PathTraversalResourceHandlerTest.java]: Something is wrong with formatting. We use tab or 4 characters for indentation.

File: pippo-core/src/main/java/ro/pippo/core/route/ClasspathResourceHandler.java [hunk 4]
```diff
@@ -32,10 +38,20 @@ public ClasspathResourceHandler(String urlPath, String resourceBasePath) {
...
+    private static Path ROOT_PATH = Paths.get("/");
+    private static String normalizePath(String subPath) {
+        return ROOT_PATH.relativize(ROOT_PATH.resolve(subPath).normalize()).toString();
+    }
+
     @Override
     public URL getResourceUrl(String resourcePath) {
-        String resourceName = getResourceBasePath() + "/" + resourcePath;
-        return this.getClass().getClassLoader().getResource(resourceName);
+        String normalizedPath = normalizePath(resourcePath);
```
Comment (decebals) [inline pippo-core/src/main/java/ro/pippo/core/route/ClasspathResourceHandler.java]: Is it not much simple to have a method:
```
// we can choose other name, if you feel that actual name is not suggestive
boolean validateResourcePath(String root, String path) {
    // put the logic here
}
```

and use this method in both `ClasspathResourceHandler` and `WebjarsResourceHandler`?

Comment (lambdaupb) [inline pippo-core/src/main/java/ro/pippo/core/route/ClasspathResourceHandler.java]: Sure, but I wont touch the `WebjarsResourceHandler`, I don't understand how it should work and don't really have the time to dive in. Feel free to change anything or not use this pull request.

**Comment by decebals** [comment]
Fixed with last commit.
@lambdaupb I appreciate very much your contribution. Thanks!

**Pull request closed without merging**